File: sw/microhil_upy/apps/microhil/commands/parser.py
# -*- coding: UTF-8 -*-

'''
Module
    parser.py
Copyright
    Copyright (C) 2026 Vladimir Roncevic <elektron.ronca@gmail.com>
    microhil is free software: you can redistribute it and/or modify it
    under the terms of the GNU General Public License as published by the
    Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.
    microhil is distributed in the hope that it will be useful, but
    WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
    See the GNU General Public License for more details.
    You should have received a copy of the GNU General Public License along
    with this program. If not, see <http://www.gnu.org/licenses/>.
Info
    Command parser of microhil.
'''

import sys
import uselect
from micropython import const

_STATE_IDLE = const(0)
_STATE_RECEIVING = const(1)

class CommandParser:
    '''
        Defines control of a command parser.

        It defines:

            :methods:
                | __init__ - Initializes the command parser.
                | process_char - Processes a character.
                | get_command - Polls stdin and returns parsed command.
    '''
    def __init__(self):
        '''
            Initializes the command parser.

            :exceptions: None.
        '''
        self.state = _STATE_IDLE
        self.buf = bytearray(64)
        self.buf_idx = 0
        self.poll = uselect.poll()
        self.poll.register(sys.stdin, uselect.POLLIN)

    def process_char(self, c):
        '''
            Processes a character.

            :param c: Character to process.
            :exceptions: None.
        '''
        if c == '\r' or c == '\n':
            return None

        if c == '<':
            self.state = _STATE_RECEIVING
            self.buf_idx = 0
            return None

        if self.state == _STATE_RECEIVING:
            if c == '>':
                cmd = self.buf[:self.buf_idx].decode()
                self.state = _STATE_IDLE
                return cmd

            if self.buf_idx < 64:
                self.buf[self.buf_idx] = ord(c)
                self.buf_idx += 1

        return None

    def get_command(self):
        '''
            Polls stdin and reads characters until a command is parsed.

            :exceptions: None.
        '''
        while self.poll.poll(0):
            c = sys.stdin.read(1)
            cmd = self.process_char(c)

            if cmd is not None:
                return cmd

        return None

File: sw/microhil_base/src/command/parser.c
/* -*- Mode: C; indent-tabs-mode: t; c-basic-offset: 4; tab-width: 4 -*-  */
/*
 * parser.c
 * Copyright (C) 2025 Vladimir Roncevic <elektron.ronca@gmail.com>
 *
 * microhil-base is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * microhil-base is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program_name. If not, see <http://www.gnu.org/licenses/>.
 */
#include "parser.h"
#include "serial/serial_transport.h"
#include <stdint.h>

typedef enum { PARSER_STATE_IDLE, PARSER_STATE_RECEIVING } parser_state_t;

static parser_state_t parser_state = PARSER_STATE_IDLE;
static uint32_t parser_idx = 0;

////////////////////////////////////////////////////////////////////////////
/// @brief Processes a single character inside the parsing state machine
///
/// @param c [in] Character to process
/// @param buf [out] Destination buffer to store parsed command
/// @param max_len [in] Maximum capacity of the destination buffer
/// @return True if a complete command frame was successfully parsed, else false
static bool parser_process_char(char c, char *buf, uint32_t max_len) {
  if (c == '\r' || c == '\n') {
    return false;
  }

  if (c == '<') {
    parser_state = PARSER_STATE_RECEIVING;
    parser_idx = 0;
    return false;
  }

  if (parser_state == PARSER_STATE_RECEIVING) {
    if (c == '>') {
      buf[parser_idx] = '\0';
      parser_state = PARSER_STATE_IDLE;
      parser_idx = 0;
      return true;
    }

    if (parser_idx < (max_len - 1)) {
      buf[parser_idx++] = c;
    } else {
      parser_state = PARSER_STATE_IDLE;
      parser_idx = 0;
    }
  }

  return false;
}

////////////////////////////////////////////////////////////////////////////
/// @brief Reads and parses command requests from the serial input stream
///
/// @param buf [out] Destination buffer to store parsed command
/// @param max_len [in] Maximum capacity of the destination buffer
/// @return True if a command was parsed, else false
bool parser_get_command(char *buf, uint32_t max_len) {
  uint8_t byte = 0;
  while (serial_transport_read_byte(&byte)) {
    if (parser_process_char((char)byte, buf, max_len)) {
      return true;
    }
  }
  return false;
}
